src/mongocxx/test/spec/unified_tests/runner.cpp

```diff
@@ -445,18 +454,219 @@ void add_data_to_collection(const array::element& data) {
     auto coll = db.create_collection(coll_name, {}, wc);
 
     auto to_insert = array_elements_to_documents(data["documents"].get_array().value);
-
-    CAPTURE(db_name, coll_name, to_insert);
-    if (!to_insert.empty())
-        REQUIRE(coll.insert_many(to_insert)->result().inserted_count() != 0);
+    return to_insert.empty() || coll.insert_many(to_insert)->result().inserted_count() != 0;
 }
 
 void load_initial_data(document::view test) {
     if (!test["initialData"])
         return;
 
-    for (const auto& datum : test["initialData"].get_array().value)
-        add_data_to_collection(datum);
+    auto data = test["initialData"].get_array().value;
+    std::all_of(std::begin(data), std::end(data), add_data_to_collection);
+}
+
+void assert_result(const array::element& ops, document::view actual_result) {
+    if (!ops["expectResult"])
+        return;
+
+    CAPTURE(to_json(actual_result));
+    auto result = ops["expectResult"];
+    test_util::assert_matches(actual_result["result"].get_value(), result.get_value());
+
+    if (ops["saveResultAsEntity"]) {
+        auto key = ops["saveResultAsEntity"].get_string().value.to_string();
+        get_entity_map().insert(key, actual_result);
+    }
+}
+
+void assert_error(const mongocxx::operation_exception& e,
```

Comment:
Operation exceptions (server errors) contain the raw server error as well (usually). This allows different logic from all other exceptions.